### DISCUSSION TITLE
Changes to how Brains Work

### DISCUSSION THREAD

**Discussion by MGD4375** [MaxRobinsonTheGreat/EvolutionSimulatorV2#87] [Ideas and Suggestions]
Currently, an organism has a brain if it has a move cell and an eye cell. We could aim to make this behaviour of linking _move_ and _eye_ generic.

![image](https://user-images.githubusercontent.com/12746100/147107032-1f39824b-9438-42aa-ae91-4c8cff7686a4.png)

**Is your idea already posted here? Don't make duplicates.**
There isn't a comparable idea being discussed.

**How exactly does your idea work? Be thorough and clear.**
1. Every cell has an input trigger, if an input trigger is fired the cell does what it's supposed to do, e.g. move, eat produce etc. 
2. Specific cell types have an output allowing them to fire when their condition is met. E.g. eye sees, producer produces etc
3. Inputs and outputs can be linked by evolution.

![image](https://user-images.githubusercontent.com/12746100/147107914-4693d43c-72d7-4ed3-a186-7b9c1ddb5fbd.png)

**How would it add to the simulation? How would it change evolution?**
The goal is to push the behaviour currently hidden in the brain of the creature into something we can draw. We may not be able to draw this in the world, but we could visualise creatures in the editor with their "neurons". 
By generalising the way brains, move cells and eye cells work, we can also expect this to open up more options of interesting cell types and combinations.

**Comment by xkortex** [comment]
I've also been thinking about better ways to evolve the brain logic. There's been [a mod which uses a neural network based brain](https://bboettcher3.github.io/LifeEngine/), but Max alluded to in the most recent video that he wants to keep things pretty simple, which adds resistance to adopting it upstream. Also neural nets are only manually editable up to a point. 

I think a decent compromise would be some sort of IFTTT (if this then that) style system. Observations are already first-class objects in the engine. So the brain could comprise a set of 0 or more "rules" which operate on observations in a Reactor pattern sort of way to produce actions.

What this would look like from a UI perspective is a set of drop-downs where you can select an observation cell type, and some sort of output. You'd want to be able to stack multiple observations in a way that composes, so I'm thinking something like: 

```
cell kind  | move affinity | rotation affinity|  clear
|v| armor  | -1            | 0                | x
|v| food   | +1            | 1                | x
|__ add row __|
```

So basically every brain cycle would consist of taking the observation stack and calculating the weight for movement and rotation, which gets converted to some action. 

I think this would be just about the simplest way to get more complex behavior that is evolve-able but also tinker-able.

**Reply by doo158** [reply]
this is cool, but as of now, i think they need to add more functional cells for it to be practical. Currently the only cells in the game that could possibly be influenced by this are the mover and producer cells
